### PR TITLE
fix(eslint-plugin): [no-unused-expressions] false negatives when using assertions

### LINKED ISSUE #8654: Bug: [no-unused-expressions] False negatives when using assertions.
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&showAST=types&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPCgK4C2ARiDPAD7zEqjYCWKIwA3AFBe5IfwA9IPhUYSagAoQADwAO4DOwCUPPvChoNKAJ4DhRJKJjiYXADxRdAPj4ChIlEbETeeAIT2DT46a5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAm9AB7FoKZJQ5NUGSImjQO0SODABfEOqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
declare const foo: number | undefined;

foo; // error (expected)

foo as any; // no error
<any>foo;  // no error
foo!;  // no error
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-expressions": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

4 errors

### Actual Result

one error

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by up201304504**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8654
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

The no-unused-expressions rule was producing false negatives when assertions are used within TypeScript code:

- TSAsExpression
- TSNonNullExpression
- TSTypeAssertion

This pull request resolves the identified bug by correcting the behavior of the no-unused-expressions rule. The changes made in this PR have been validated in the TypeScript playground:

fixed_false_negatives_unused_expressions

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts [hunk 1]
```diff
@@ -325,5 +325,49 @@ Foo<string>;
...
+    {
+      code: 'declare const foo: number | undefined; foo as any;',
+      errors: [
+        {
+          line: 1,
+          endLine: 1,
+          column: 40,
+          endColumn: 51,
+        },
+      ],
```
Comment (auvred) [inline packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts:348]:
```suggestion
      errors: error([
        {
          line: 1,
          endLine: 1,
          column: 40,
          endColumn: 51,
        },
      ]),
```

[Testing] Other `invalid` test cases use this `error` helper function. Let's use it for all newly added cases for better consistency?

**Review by auvred — CHANGES_REQUESTED** [review]
Nice start! :rocket: 

The code changes look good to me. I'm requesting one change in tests.

File: packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts (hunk 1 above)
Comment (up201304504) [reply packages/eslint-plugin/tests/rules/no-unused-expressions.test.ts:348]: Hi! Thanks for the feedack! I'm going to take a look :)

Additionally, I just realised that my pipeline is failing on `lint with build`? Any command to run it just for the classes w/changes?

**Review by auvred — APPROVED** [review]
:rocket: LGTM! Thanks for working on this!

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥 thanks!

The word 'FIRE' animated to have flames suddenly coming out of it. The 'F' is an orange-ish molten rock pattern.

**Pull request merged**